contracts/pcv/PCVGuardian.sol

```diff
@@ -0,0 +1,245 @@
+import {IPCVGuardian} from "./IPCVGuardian.sol";
+import {IPCVDeposit} from "./IPCVDeposit.sol";
+import {CoreRefPauseableLib} from "../libs/CoreRefPauseableLib.sol";
+import {TribeRoles} from "../core/TribeRoles.sol";
+
+/// @notice PCV Guardian is a contract to safeguard protocol funds
+/// by being able to withdraw whitelisted PCV deposits to an immutable safe address
+contract PCVGuardian is IPCVGuardian, CoreRef {
+    using CoreRefPauseableLib for address;
+    using EnumerableSet for EnumerableSet.AddressSet;
+
+    ///@notice set of whitelisted pcvDeposit addresses for withdrawal
+    EnumerableSet.AddressSet private whitelistAddresses;
+
+    ///@notice safe address where PCV funds can be withdrawn to
+    address public immutable safeAddress;
+
+    constructor(
+        address _core,
+        address _safeAddress,
+        address[] memory _whitelistAddresses
+    ) CoreRef(_core) {
+        safeAddress = _safeAddress;
+
+        // improbable to ever overflow
+        unchecked {
+            for (uint256 i = 0; i < _whitelistAddresses.length; i++) {
+                _addWhitelistAddress(_whitelistAddresses[i]);
+            }
+        }
+    }
+
+    /// @dev checks if a pcv deposit address is whitelisted, reverts if not
+    modifier onlyWhitelist(address pcvDeposit) {
+        require(
+            isWhitelistAddress(pcvDeposit),
+            "PCVGuardian: Provided address is not whitelisted"
+        );
+        _;
+    }
+
+    // ---------- Read-Only API ----------
+
+    /// @notice returns true if the the provided address is a valid destination to withdraw funds from
+    /// @param pcvDeposit the pcvDeposit address to check if whitelisted
+    function isWhitelistAddress(address pcvDeposit)
+        public
+        view
+        override
+        returns (bool)
+    {
+        return whitelistAddresses.contains(pcvDeposit);
+    }
+
+    /// @notice returns all whitelisted pcvDeposit addresses
+    function getWhitelistAddresses()
+        public
+        view
+        override
+        returns (address[] memory)
+    {
+        return whitelistAddresses.values();
+    }
+
+    // ---------- Governor-Only State-Changing API ----------
+
+    /// @notice governor-only method to whitelist a pcvDeposit address to withdraw funds from
+    /// @param pcvDeposit the address to whitelist
+    function addWhitelistAddress(address pcvDeposit)
+        external
+        override
+        onlyGovernor
+    {
+        _addWhitelistAddress(pcvDeposit);
+    }
+
+    /// @notice batch version of addWhiteListaddress
+    /// @param _whitelistAddresses the addresses to whitelist, as calldata
+    function addWhitelistAddresses(address[] calldata _whitelistAddresses)
+        external
+        override
+        onlyGovernor
+    {
+        // improbable to ever overflow
+        unchecked {
+            require(
+                _whitelistAddresses.length != 0,
+                "PCVGuardian: Empty address array provided"
```

Comment:
I don't think we need this check as the function will do nothing if the length is 0